### DISCUSSION TITLE
Determining how long a task has been enqueued

### DISCUSSION THREAD

**Discussion by lindsayweil-sl** [hibiken/asynq#523] [Q&A]
I've been looking through the documentation and not finding a way to do this, so hoping someone might have seen something that I am overlooking.  I would like to know how long a given task has been enqueued for monitoring purposes - to know if it was enqueued 2 minutes ago, for example, and report this to our monitoring tools.

I see TaskInfo has retry and failure timestamps, and opts has processAt, but I'm not seeing a way from a task to know when it was enqueued.  Is there a time component that I'm overlooking?

**Comment by lindsayweil-sl** [comment]
Currently looking at the Latency queue stats to see if they'll work for what we need

**Reply by lindsayweil-sl** [reply]
The queue latency from the Inspector seems to give me what I need.

**Reply by lindsayweil-sl** [reply]
Adding detail in case anyone else needs this:

- getting an Inspector: https://github.com/hibiken/asynq/blob/c70ff6a335eed05735427567e8e9549745f43204/inspector.go#L26
- ask Inspector for queue info: https://github.com/hibiken/asynq/blob/c70ff6a335eed05735427567e8e9549745f43204/inspector.go#L125
- get Latency from the QueueInfo

**Comment by hibiken** [comment]
@lindsayweil-sl Thanks for opening a discussion!

Asynq internals records timestamps for each task when it transitioned to "pending" state. This time should be roughly the time user specifies in `ProcessAt` (now + `ProcessIn`) if using delayed task, otherwise the `time.Now()` when user enqueues. 

Currently, we provide "Queue latency" which is the oldest pending tasks in a given queue. 

However, I think it's valuable to provide age of task in `TaskInfo` struct if the task is in pending state. Let me know your thoughts!

**Reply by lindsayweil-sl** [reply]
@hibiken I think the queue latency will give us what we need for the monitoring I was looking for as we really care about the oldest job anyway.  Having some way of capturing and accessing when the task was enqueued might be helpful just to be able to tell how old a given task is, but I think I have what I need for now!
